From 466e392b60385602173c0fcbeb2d98b63fa12d8b Mon Sep 17 00:00:00 2001
From: Johnothan King <johnothanking@protonmail.com>
Date: Sat, 28 Mar 2026 21:37:56 -0700
Subject: size_t/ptrdiff_t transition part 12: ksh93 lexing, parsing and
 subshells

This is the twelfth of the thickfold patch series, which enables ksh93
to operate within a 64-bit address space.

The parts of ksh93 affected by this commit are:
- The lexing and parsing components in lex.c, parse.c, fcin.c
  and trestore.c.
  - Added an assert() after stkalloc() to fix the following
    LTO warning:
    warning: 'memset' writing 60 bytes into a region of size 0 overflows the destination [-Wstringop-overflow=]
      1819 |                         memset(ioq,0,sizeof(*ioq));
           |                         ^
    lto1: note: destination object is likely at address zero
- Minor fixes for the virtual subshell mechanism.
- The code underlying shcomp(1), aka sh_tdump().
- A minor fix to a cast in sh_timeradd().

Change in the number of warnings on Linux when compiling with clang using
-Wsign-compare -Wshorten-64-to-32 -Wsign-conversion -Wimplicit-int-conversion:
482 => 353 => 37 (progression from part 11 => part 12 => part 13)

Progresses https://github.com/ksh93/ksh/issues/592
---
 src/cmd/ksh93/include/fcin.h    |  15 +--
 src/cmd/ksh93/include/shlex.h   |  16 +--
 src/cmd/ksh93/include/shnodes.h |   4 +-
 src/cmd/ksh93/sh/fcin.c         |  25 +++--
 src/cmd/ksh93/sh/lex.c          | 188 +++++++++++++++++---------------
 src/cmd/ksh93/sh/parse.c        |  94 ++++++++--------
 src/cmd/ksh93/sh/shcomp.c       |   2 +-
 src/cmd/ksh93/sh/subshell.c     |  12 +-
 src/cmd/ksh93/sh/tdump.c        |  38 +++----
 src/cmd/ksh93/sh/timers.c       |   4 +-
 src/cmd/ksh93/sh/trestore.c     |  52 ++++-----
 11 files changed, 235 insertions(+), 215 deletions(-)

diff --git a/src/cmd/ksh93/include/fcin.h b/src/cmd/ksh93/include/fcin.h
index ce325bbcc..22a7e7117 100644
--- a/src/cmd/ksh93/include/fcin.h
+++ b/src/cmd/ksh93/include/fcin.h
@@ -12,6 +12,7 @@
 *                                                                      *
 *                  David Korn <dgk@research.att.com>                   *
 *                  Martijn Dekker <martijn@inlv.org>                   *
+*            Johnothan King <johnothanking@protonmail.com>             *
 *                                                                      *
 ***********************************************************************/
 #ifndef _FCIN_H
@@ -34,9 +35,9 @@ typedef struct _fcin
 	unsigned char	*fcptr;		/* pointer to next input char */
 	unsigned char	fcchar;		/* saved character */
 	short		fclen;		/* last multibyte char len */
-	void (*fcfun)(Sfio_t*,const char*,int,void*);	/* advance function */
+	void (*fcfun)(Sfio_t*,const char*,ptrdiff_t,void*);	/* advance function */
 	void		*context;	/* context pointer */
-	int		fcleft;		/* for multibyte boundary */
+	ptrdiff_t	fcleft;		/* for multibyte boundary */
 } Fcin_t;
 
 #if SHOPT_MULTIBYTE
@@ -46,9 +47,9 @@ typedef struct _fcin
 #   define fcmbget(x)	(fcget())
 #endif
 #define fcfile()	(_Fcin._fcfile)
-#define fcgetc()	(*_Fcin.fcptr++ ? (int)_Fcin.fcptr[-1] : fcfill())
-#define fcget()		((int)(*_Fcin.fcptr++))
-#define fcpeek(n)	((int)_Fcin.fcptr[n])
+#define fcgetc()	(*_Fcin.fcptr++ ? _Fcin.fcptr[-1] : fcfill())
+#define fcget()		(*_Fcin.fcptr++)
+#define fcpeek(n)	(_Fcin.fcptr[n])
 #define fcseek(n)	((char*)(_Fcin.fcptr+=(n)))
 #define fcfirst()	((char*)_Fcin.fcbuff)
 #define fclast()	((char*)_Fcin.fclast)
@@ -57,9 +58,9 @@ typedef struct _fcin
 #define fcsave(x)	(*(x) = _Fcin)
 #define fcrestore(x)	(_Fcin = *(x))
 extern int		fcfill(void);
-extern int		fcfopen(Sfio_t*);
+extern ssize_t		fcfopen(Sfio_t*);
 extern int		fcclose(void);
-void			fcnotify(void(*)(Sfio_t*,const char*,int,void*),void*);
+void			fcnotify(void(*)(Sfio_t*,const char*,ptrdiff_t,void*),void*);
 
 extern Fcin_t		_Fcin;		/* used by macros */
 
diff --git a/src/cmd/ksh93/include/shlex.h b/src/cmd/ksh93/include/shlex.h
index 8c0dd6b10..f2ecaee0a 100644
--- a/src/cmd/ksh93/include/shlex.h
+++ b/src/cmd/ksh93/include/shlex.h
@@ -72,21 +72,21 @@ typedef struct  _shlex_
 {
 	struct argnod	*arg;		/* current word */
 	struct ionod	*heredoc;	/* pending here document list */
+	ssize_t		varnamelength;	/* length of variable name in assignment */
 	int		token;		/* current token number */
 	int		lastline;	/* last line number */
 	int		lasttok;	/* previous token number */
 	int		digits;		/* numerical value with word token */
+	int		inlineno;	/* saved value of sh.inlineno */
+	int		firstline;	/* saved value of sh.st.firstline */
+	int		assignlevel;	/* nesting level for assignment */
 	char		aliasok;	/* on when alias is legal */
 	char		assignok;	/* on when name=value is legal */
-	int		varnamelength;	/* length of variable name in assignment */
 	char		inexec;		/* on when processing exec */
 	char		intypeset;	/* 1 when processing typeset, 2 when processing enum */
 	char		comp_assign;	/* in compound assignment */
 	char		comsub;		/* parsing command substitution */
 	char		noreserv;	/* reserved words not legal */
-	int		inlineno;	/* saved value of sh.inlineno */
-	int		firstline;	/* saved value of sh.st.firstline */
-	int		assignlevel;	/* nesting level for assignment */
 	/* The following two struct members are considered private to lex.c */
 	struct _shlex_pvt_lexdata_  lexd;
 	struct _shlex_pvt_lexstate_  lex;
@@ -168,20 +168,20 @@ typedef struct  _shlex_
 extern int		sh_lex(Lex_t*);
 extern Shnode_t		*sh_dolparen(Lex_t*);
 extern Lex_t		*sh_lexopen(Lex_t*, int);
-extern void 		sh_lexskip(Lex_t*,int,int,int);
+extern void 		sh_lexskip(Lex_t*,char,int,int);
 extern noreturn void 	sh_syntax(Lex_t*, int);
 
 #if SHOPT_KIA
     typedef struct
     {
-	off_t		offset;
+	ptrdiff_t	offset;
 	Sfio_t		*file;		/* kia output file */
 	Sfio_t		*tmp;		/* kia reference file */
 	unsigned long	script;		/* script entity number */
 	unsigned long	fscript;	/* script file entity number */
 	unsigned long	current;	/* current entity number */
 	unsigned long	unknown;	/* <unknown> entity number */
-	off_t		begin;		/* offset of first entry */
+	Sfoff_t		begin;		/* offset of first entry */
 	char		*scriptname;	/* name of script file */
 	Dt_t		*entity_tree;	/* for entity IDs */
     } Kia_t;
@@ -189,7 +189,7 @@ extern noreturn void 	sh_syntax(Lex_t*, int);
     extern Kia_t		kia;
 
     extern int                  kiaclose(Lex_t *);
-    extern unsigned long        kiaentity(Lex_t*, const char*,int,int,int,int,unsigned long,int,int,const char*);
+    extern unsigned long        kiaentity(Lex_t*, const char*,ssize_t,int,int,int,unsigned long,int,size_t,const char*);
 #endif /* SHOPT_KIA */
 
 #endif /* !_SHLEX_H */
diff --git a/src/cmd/ksh93/include/shnodes.h b/src/cmd/ksh93/include/shnodes.h
index 510d1c165..2b0829af8 100644
--- a/src/cmd/ksh93/include/shnodes.h
+++ b/src/cmd/ksh93/include/shnodes.h
@@ -209,8 +209,8 @@ union Shnode_u
 
 extern void			sh_freeup(void);
 extern void			sh_funstaks(struct slnod*,int);
-extern Sfio_t 			*sh_subshell(Shnode_t*, volatile int, int);
-extern int			sh_tdump(Sfio_t*, const Shnode_t*);
+extern Sfio_t 			*sh_subshell(Shnode_t*, volatile int, char);
+extern ssize_t			sh_tdump(Sfio_t*, const Shnode_t*);
 extern Shnode_t			*sh_trestore(Sfio_t*);
 
 #endif /* !_SHNODES_H */
diff --git a/src/cmd/ksh93/sh/fcin.c b/src/cmd/ksh93/sh/fcin.c
index 2badaa5b0..5fc2984c7 100644
--- a/src/cmd/ksh93/sh/fcin.c
+++ b/src/cmd/ksh93/sh/fcin.c
@@ -2,7 +2,7 @@
 *                                                                      *
 *               This software is part of the ast package               *
 *          Copyright (c) 1982-2011 AT&T Intellectual Property          *
-*          Copyright (c) 2020-2024 Contributors to ksh 93u+m           *
+*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
 *                      and is licensed under the                       *
 *                 Eclipse Public License, Version 2.0                  *
 *                                                                      *
@@ -34,9 +34,9 @@ Fcin_t _Fcin = {0};
 /*
  * open stream <f> for fast character input
  */
-int	fcfopen(Sfio_t* f)
+ssize_t	fcfopen(Sfio_t* f)
 {
-	int	n;
+	ssize_t	n;
 	char	*buff;
 	Fcin_t	save;
 	errno = 0;
@@ -70,9 +70,9 @@ int	fcfopen(Sfio_t* f)
  * If last is non-zero, and the stream is a file, 0 is returned when
  * the previous character is a 0 byte.
  */
-int	fcfill(void)
+int fcfill(void)
 {
-	int	n;
+	ptrdiff_t n;
 	Sfio_t	*f;
 	unsigned char	*last=_Fcin.fclast, *ptr=_Fcin.fcptr;
 	if(!(f=fcfile()))
@@ -93,7 +93,7 @@ int	fcfill(void)
 	}
 	if((n = ptr-_Fcin.fcbuff) && _Fcin.fcfun)
 		(*_Fcin.fcfun)(f,(const char*)_Fcin.fcbuff,n,_Fcin.context);
-	sfread(f, (char*)_Fcin.fcbuff, n);
+	sfread(f, (char*)_Fcin.fcbuff, (size_t)n);
 	_Fcin._fcfile = 0;
 	if(!last)
 		return 0;
@@ -122,7 +122,7 @@ int fcclose(void)
 /*
  * Set the notify function that is called for each fcfill()
  */
-void fcnotify(void (*fun)(Sfio_t*,const char*,int,void*),void* context)
+void fcnotify(void (*fun)(Sfio_t*,const char*,ptrdiff_t,void*),void* context)
 {
 	_Fcin.fcfun = fun;
 	_Fcin.context = context;
@@ -151,7 +151,8 @@ struct Extra
 int _fcmbget(short *len)
 {
 	static struct Extra	extra;
-	int			i, c, n;
+	int			c;
+	ptrdiff_t		i, n;
 	/*
 	 * Check if we need to piece together a split multibyte
 	 * character started at the end of the previous buffer.
@@ -165,7 +166,7 @@ int _fcmbget(short *len)
 			_Fcin.fcptr = (unsigned char*)fcfirst() - _Fcin.fcleft; 
 			_Fcin.fcleft = 0;
 		}
-		*len = c;
+		*len = (short)c;
 		if(c==1)
 			c = *extra.next++;
 		else if(c==0)
@@ -174,7 +175,7 @@ int _fcmbget(short *len)
 			c = mbchar(extra.next);
 		return c;
 	}
-	switch(*len = mbsize(_Fcin.fcptr))
+	switch(*len = (short)mbsize(_Fcin.fcptr))
 	{
 	    case -1:
 		/*
@@ -183,11 +184,11 @@ int _fcmbget(short *len)
 		 */
 		if(_Fcin._fcfile && (n=(_Fcin.fclast-_Fcin.fcptr)) < MB_LEN_MAX)
 		{
-			memcpy(extra.buff, _Fcin.fcptr, n);
+			memcpy(extra.buff, _Fcin.fcptr, (size_t)n);
 			_Fcin.fcptr = _Fcin.fclast;
 			/* fcgetc() will read the next buffer via fcfill() */
 			for(i=n; i < MB_LEN_MAX+n; i++)
-				if((extra.buff[i] = fcgetc())==0)
+				if((extra.buff[i] = (unsigned char)fcgetc())==0)
 					break;
 			_Fcin.fcleft = n;
 			extra.next = extra.buff;
diff --git a/src/cmd/ksh93/sh/lex.c b/src/cmd/ksh93/sh/lex.c
index 1160c57f3..79ef35eb4 100644
--- a/src/cmd/ksh93/sh/lex.c
+++ b/src/cmd/ksh93/sh/lex.c
@@ -73,7 +73,7 @@ static int		alias_exceptf(Sfio_t*, int, void*, Sfdisc_t*);
 static void		setupalias(Lex_t*,const char*, Namval_t*);
 static int		comsub(Lex_t*,int);
 static void		nested_here(Lex_t*);
-static int		here_copy(Lex_t*, struct ionod*);
+static ptrdiff_t	here_copy(Lex_t*, struct ionod*);
 static int 		stack_grow(void);
 static const Sfdisc_t alias_disc = { NULL, NULL, NULL, alias_exceptf, NULL };
 
@@ -84,16 +84,17 @@ static int		lex_max, *lex_match;
 
 static void refvar(Lex_t *lp, int type)
 {
-	off_t off = (fcseek(0)-(type+1))-(lp->lexd.first?lp->lexd.first:fcfirst());
+	ptrdiff_t off = (fcseek(0)-(type+1))-(lp->lexd.first?lp->lexd.first:fcfirst());
 	unsigned long r;
 	if(lp->lexd.first)
 	{
 		off = (fcseek(0)-(type+1)) - lp->lexd.first;
-		r=kiaentity(lp,lp->lexd.first+kia.offset+type,off-kia.offset,'v',-1,-1,kia.current,'v',0,"");
+		r=kiaentity(lp,lp->lexd.first+kia.offset+type,(ssize_t)(off-kia.offset),'v',-1,-1,kia.current,'v',0,"");
 	}
 	else
 	{
-		int n,offset = stktell(sh.stk);
+		ptrdiff_t offset = stktell(sh.stk);
+		ssize_t n;
 		void *savptr;
 		char *begin;
 		off = offset + (fcseek(0)-(type+1)) - fcfirst();
@@ -101,15 +102,15 @@ static void refvar(Lex_t *lp, int type)
 		{
 			/* variable starts on stack, copy remainder */
 			if(off>offset)
-				sfwrite(sh.stk,fcfirst()+type,off-offset);
-			n = stktell(sh.stk)-kia.offset;
+				sfwrite(sh.stk,fcfirst()+type,(size_t)(off-offset));
+			n = (ssize_t)(stktell(sh.stk)-kia.offset);
 			begin = stkptr(sh.stk,kia.offset);
 		}
 		else
 		{
 			/* variable in data buffer */
 			begin = fcfirst()+(type+kia.offset-offset);
-			n = off-kia.offset;
+			n = (ssize_t)(off-kia.offset);
 		}
 		savptr = stkfreeze(sh.stk,0);
 		r=kiaentity(lp,begin,n,'v',-1,-1,kia.current,'v',0,"");
@@ -123,23 +124,23 @@ static void refvar(Lex_t *lp, int type)
  * This routine gets called when reading across a buffer boundary
  * If lexd.nocopy is off, then current token is saved on the stack
  */
-static void lex_advance(Sfio_t *iop, const char *buff, int size, void *context)
+static void lex_advance(Sfio_t *iop, const char *buff, ptrdiff_t size, void *context)
 {
 	Lex_t		*lp = (Lex_t*)context;
 	/* write to history file and to stderr if necessary */
 	if(iop && !sfstacked(iop))
 	{
 		if(sh_isstate(SH_HISTORY) && sh.hist_ptr)
-			sfwrite(sh.hist_ptr->histfp, buff, size);
+			sfwrite(sh.hist_ptr->histfp, buff, (size_t)size);
 		if(sh_isstate(SH_VERBOSE))
-			sfwrite(sfstderr, buff, size);
+			sfwrite(sfstderr, buff, (size_t)size);
 	}
 	if(lp->lexd.nocopy)
 		return;
 	if(lp->lexd.dolparen && lp->lexd.docword && lp->lexd.docend)
 	{
-		int n = size - (lp->lexd.docend-(char*)buff);
-		sfwrite(sh.strbuf,lp->lexd.docend,n);
+		ptrdiff_t n = size - (lp->lexd.docend-(char*)buff);
+		sfwrite(sh.strbuf,lp->lexd.docend,(size_t)n);
 		lp->lexd.docextra  += n;
 		if(sffileno(iop)>=0)
 			lp->lexd.docend = sfsetbuf(iop,iop,0);
@@ -148,7 +149,7 @@ static void lex_advance(Sfio_t *iop, const char *buff, int size, void *context)
 	}
 	if(lp->lexd.first)
 	{
-		size -= (lp->lexd.first-(char*)buff);
+		size -= lp->lexd.first-(char*)buff;
 		buff = lp->lexd.first;
 		if(!lp->lexd.inlexskip)
 			lp->arg = stkseek(sh.stk,ARGVAL);
@@ -158,7 +159,7 @@ static void lex_advance(Sfio_t *iop, const char *buff, int size, void *context)
 	}
 	if(size>0 && (lp->arg||lp->lexd.inlexskip))
 	{
-		sfwrite(sh.stk,buff,size);
+		sfwrite(sh.stk,buff,(size_t)size);
 		lp->lexd.first = 0;
 	}
 }
@@ -172,7 +173,8 @@ static int lexfill(Lex_t *lp)
 	int c;
 	Lex_t savelex;
 	struct argnod *ap;
-	int aok,docextra;
+	int docextra;
+	char aok;
 	savelex = *lp;
 	ap = lp->arg;
 	c = fcfill();
@@ -254,10 +256,12 @@ int sh_lex(Lex_t *lp)
 int sh_lex(Lex_t* lp)
 {
 	const char	*state;
-	int		n, c, mode=ST_BEGIN, wordflags=0;
+	ptrdiff_t	n;
+	int		c, mode=ST_BEGIN;
+	uint8_t		argflgs, wordflags=0;
 	int		inlevel=lp->lexd.level, assignment=0, ingrave=0;
 	int		epatchar=0;
-	int		varnametry = 0, varnamecount = 0, varnamelength = 0;
+	ssize_t		varnametry = 0, varnamecount = 0, varnamelength = 0;
 	SETLEN(1);
 	if(lp->lexd.paren)
 	{
@@ -407,7 +411,7 @@ int sh_lex(Lex_t* lp)
 						fcseek(1);
 				}
 				while(c=='#');
-				lp->lexd.nocopy = n;
+				lp->lexd.nocopy = (char)n;
 				if(c<0)
 					return lp->token=EOFSYM;
 				n = S_NLTOK;
@@ -637,9 +641,9 @@ int sh_lex(Lex_t* lp)
 						fcsopen((char*)state);
 					/* remove \new-line */
 					n = stktell(sh.stk)-c;
-					stkseek(sh.stk,n);
+					stkseek(sh.stk,(ptrdiff_t)n);
 					lp->arg = ap;
-					if(n<=ARGVAL)
+					if(n<=(ssize_t)ARGVAL)
 					{
 						mode = 0;
 						lp->lexd.first = 0;
@@ -807,7 +811,7 @@ int sh_lex(Lex_t* lp)
 				else if((n=endchar(lp))==c)
 				{
 					if(sh.inlineno > lp->lastline)
-						lp->lex.last_quote = c;
+						lp->lex.last_quote = (char)c;
 					/*
 					 * At this point, we know that the previous skipping of characters was done
 					 * according to the ST_QUOTE state table. We also know that the character that
@@ -846,7 +850,7 @@ int sh_lex(Lex_t* lp)
 				fcseek(-LEN);
 				n = lp->digits;
 				wordflags |= comsub(lp,c);
-				lp->digits = n;
+				lp->digits = (int)n;
 				continue;
 			case S_RBRA:
 				if((n=endchar(lp)) == '$')
@@ -880,7 +884,7 @@ int sh_lex(Lex_t* lp)
 					if(n>0)
 						fcseek(-LEN);
 				}
-				if(isaletter(n) || n==LBRACT)
+				if(isaletter((wchar_t)n) || n==LBRACT)
 					continue;
 				if(mode==ST_NAME)
 				{
@@ -991,7 +995,7 @@ int sh_lex(Lex_t* lp)
 						{
 							if(c!='%')
 							{
-								lp->token = n;
+								lp->token = (int)n;
 								sh_syntax(lp,0);
 							}
 							else if(lp->lexd.warn)
@@ -1275,11 +1279,11 @@ breakloop:
 	if(!lp->arg)
 		lp->arg = stkseek(sh.stk,ARGVAL);
 	if(n>0)
-		sfwrite(sh.stk,state,n);
+		sfwrite(sh.stk,state,(size_t)n);
 	sfputc(sh.stk,0);
 	stkseek(sh.stk,stktell(sh.stk)-1);
 	state = stkptr(sh.stk,ARGVAL);
-	n = stktell(sh.stk)-ARGVAL;
+	n = stktell(sh.stk)-(ptrdiff_t)ARGVAL;
 	lp->lexd.first=0;
 	if(n==1)
 	{
@@ -1288,17 +1292,17 @@ breakloop:
 		if(!lp->lex.intest && (c=='<' || c=='>') && isadigit(n))
 		{
 			c = sh_lex(lp);
-			lp->digits = (n-'0');
+			lp->digits = ((int)n-'0');
 			return c;
 		}
 		if(n==LBRACT)
-			c = 0;
+			argflgs = 0;
 		else if(n==RBRACE && lp->comsub)
-			return lp->token=n;
+			return lp->token=(int)n;
 		else if(n=='~')
-			c = ARG_MAC;
+			argflgs = ARG_MAC;
 		else
-			c = (wordflags&ARG_EXP);
+			argflgs = (wordflags&ARG_EXP);
 		n = 1;
 	}
 	else if(n>2 && state[0]=='{' && state[n-1]=='}' && !lp->lex.intest && !lp->lex.incase && (c=='<' || c== '>'))
@@ -1310,43 +1314,43 @@ breakloop:
 			lp->arg = stkfreeze(sh.stk,1);
 			return lp->token=IOVNAME;
 		}
-		c = wordflags;
+		argflgs = wordflags;
 	}
 	else
-		c = wordflags;
+		argflgs = wordflags;
 	if(assignment || (lp->lex.intest&&!lp->lex.incase) || mode==ST_NONE)
-		c &= ~ARG_EXP;
-	if((c&ARG_EXP) && (c&ARG_QUOTED))
-		c |= ARG_MAC;
+		argflgs &= ~ARG_EXP;
+	if((argflgs&ARG_EXP) && (argflgs&ARG_QUOTED))
+		argflgs |= ARG_MAC;
 	if(mode==ST_NONE)
 	{
 		/* eliminate trailing )) */
 		stkseek(sh.stk,stktell(sh.stk)-2);
 	}
-	if(c&ARG_MESSAGE)
+	if(argflgs&ARG_MESSAGE)
 	{
 		if(sh_isoption(SH_DICTIONARY))
 			lp->arg = endword(2);
-		c |= ARG_MAC;
+		argflgs |= ARG_MAC;
 	}
-	if(c==0 || (c&(ARG_MAC|ARG_EXP|ARG_MESSAGE)))
+	if(argflgs==0 || (argflgs&(ARG_MAC|ARG_EXP|ARG_MESSAGE)))
 	{
 		lp->arg = stkfreeze(sh.stk,1);
-		lp->arg->argflag = (c?c:ARG_RAW);
+		lp->arg->argflag = (argflgs?argflgs:ARG_RAW);
 	}
 	else if(mode==ST_NONE)
 		lp->arg = endword(-1);
 	else
 		lp->arg = endword(0);
 	state = lp->arg->argval;
-	lp->comp_assign = assignment;
+	lp->comp_assign = (char)assignment;
 	if(assignment)
 	{
 		lp->arg->argflag |= ARG_ASSIGN;
 		lp->varnamelength = varnamelength;
 		if(sh_isoption(SH_NOEXEC))
 		{
-			char *cp = strchr(state, '=');
+			char *cp = (char*)strchr(state, '=');
 			if(cp && strncmp(++cp, "$((", 3) == 0)
 				errormsg(SH_DICT, ERROR_warn(0), e_lexarithwarn, sh.inlineno,
 					state, cp - state, state, cp + 3);
@@ -1392,7 +1396,7 @@ breakloop:
 			lp->token = ETESTSYM;
 			return lp->token;
 		}
-		c = sh_lookup(state,shtab_testops);
+		c = (int)sh_lookup(state,shtab_testops);
 		switch(c)
 		{
 		case TEST_SEQ:
@@ -1474,7 +1478,7 @@ breakloop:
 	if(!lp->lex.skipword)
 	{
 		if(n>1 && lp->lex.reservok==1 && mode==ST_NAME &&
-			(c=sh_lookup(state,shtab_reserved)))
+			(c=(int)sh_lookup(state,shtab_reserved)))
 		{
 			if(lp->lex.incase)
 			{
@@ -1556,15 +1560,17 @@ breakloop:
  */
 static int comsub(Lex_t *lp, int endtok)
 {
-	int n,c;
+	ptrdiff_t n;
+	int c;
 	unsigned short count=1;
 	int line=sh.inlineno;
 	struct ionod *inheredoc = lp->heredoc;
 	char save_arithexp = lp->lexd.dolparen_arithexp;
 	char *first,*cp=fcseek(0),word[5];
-	int off, messages=0, assignok=lp->assignok, csub;
+	int messages=0;
+	char assignok=lp->assignok, csub=lp->comsub;
+	ptrdiff_t off;
 	struct _shlex_pvt_lexstate_ save = lp->lex;
-	csub = lp->comsub;
 	sh_lexopen(lp,1);
 	lp->lexd.dolparen++;
 	lp->lexd.dolparen_arithexp = endtok==LPAREN && fcpeek(1)==LPAREN;  /* $(( */
@@ -1579,7 +1585,7 @@ static int comsub(Lex_t *lp, int endtok)
 		c=*cp, *cp=0;
 	n = sh_lex(lp);
 	if(off<0)
-		*cp = c;
+		*cp = (char)c;
 	if(n==endtok || off<0)
 	{
 		if(endtok==LPAREN && lp->lexd.paren)
@@ -1614,7 +1620,7 @@ static int comsub(Lex_t *lp, int endtok)
 					}
 					goto skip;
 				}
-				word[n++] = c;
+				word[n++] = (char)c;
 			}
 			if(sh_lexstates[ST_NAME][c]==S_BREAK)
 			{
@@ -1714,14 +1720,15 @@ done:
 static void nested_here(Lex_t *lp)
 {
 	struct ionod	*iop;
-	int		n=0,offset;
+	size_t		n=0;
+	ptrdiff_t	offset;
 	struct argnod	*arg = lp->arg;
 	char		*base;
 	if(offset=stktell(sh.stk))
 		base = stkfreeze(sh.stk,0);
 	if(lp->lexd.docend)
-		n = fcseek(0)-lp->lexd.docend;
-	iop = sh_newof(0,struct ionod,1,lp->lexd.docextra+n+ARGVAL);
+		n = (size_t)(fcseek(0)-lp->lexd.docend);
+	iop = sh_newof(0,struct ionod,1,(size_t)lp->lexd.docextra+n+ARGVAL);
 	iop->iolst = lp->heredoc;
 	stkseek(sh.stk,ARGVAL);
 	if(lp->lexd.docextra)
@@ -1751,7 +1758,7 @@ static void nested_here(Lex_t *lp)
  * if <copy> is non,zero, then the characters are copied to the stack
  * <state> is the initial lexical state
  */
-void sh_lexskip(Lex_t *lp,int close, int copy, int  state)
+void sh_lexskip(Lex_t *lp, char close, int copy, int state)
 {
 	char	*cp;
 	lp->lexd.nest = close;
@@ -1765,11 +1772,12 @@ void sh_lexskip(Lex_t *lp,int close, int copy, int  state)
 	lp->lexd.inlexskip = 0;
 	if(copy)
 	{
+		ptrdiff_t len;
 		fcnotify(0,lp);
 		if(!(cp=lp->lexd.first))
 			cp = fcfirst();
-		if((copy = fcseek(0)-cp) > 0)
-			sfwrite(sh.stk,cp,copy);
+		if((len = fcseek(0)-cp) > 0)
+			sfwrite(sh.stk,cp,(size_t)len);
 	}
 	else
 		lp->lexd.nocopy--;
@@ -1779,21 +1787,21 @@ void sh_lexskip(Lex_t *lp,int close, int copy, int  state)
     ssize_t _sfwrite(Sfio_t *sp, const void *buff, size_t n)
     {
 	const char *cp = (const char*)buff, *next=cp, *ep = cp + n;
-	int m=0,k;
-	while(next = (const char*)memchr(next,'\r',ep-next))
+	ssize_t m=0,k;
+	while(next = (const char*)memchr(next,'\r',(size_t)(ep-next)))
 		if(*++next=='\n')
 		{
 			if(k=next-cp-1)
 			{
-				if((k=sfwrite(sp,cp,k)) < 0)
+				if((k=sfwrite(sp,cp,(size_t)k)) < 0)
 					return m>0?m:-1;
 				m += k;
 			}
 			cp = next;
 		}
-	if((k=sfwrite(sp,cp,ep-cp)) < 0)
+	if((k=sfwrite(sp,cp,(size_t)(ep-cp))) < 0)
 		return m>0?m:-1;
-	return m+k;
+	return (ssize_t)(m+k);
     }
 #   define sfwrite	_sfwrite
 #endif /* SHOPT_CRNL */
@@ -1804,13 +1812,13 @@ void sh_lexskip(Lex_t *lp,int close, int copy, int  state)
  * noted with the IOQUOTE flag
  * returns 1 for complete here-doc, 0 for EOF
  */
-static int here_copy(Lex_t *lp,struct ionod *iop)
+static ptrdiff_t here_copy(Lex_t *lp,struct ionod *iop)
 {
 	const char	*state;
-	int		c,n;
+	ptrdiff_t	c,n, nsave;
 	char		*bufp,*cp;
 	Sfio_t		*sp=sh.heredocs;
-	int		stripcol=0,stripflg, nsave, special=0;
+	int		stripcol=0,stripflg, special=0;
 	if(iop->iolst)
 		here_copy(lp,iop->iolst);
 	iop->iooffset = sfseek(sp,0,SEEK_END);
@@ -1824,7 +1832,7 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 		/* skip over leading tabs in document */
 		if(iop->iofile&IOLSEEK)
 		{
-			iop->iofile &= ~IOLSEEK;
+			iop->iofile &= (unsigned)~IOLSEEK;
 			while((c = fcgetc())=='\t' || c==' ')
 			{
 				if(c==' ')
@@ -1854,7 +1862,7 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 				if(mbsize(fcseek(0)) < 0 && fcleft() < MB_LEN_MAX)
 				{
 					n = S_EOF;
-					SETLEN(-fcleft());
+					SETLEN((short)-fcleft());
 					break;
 				}
 			}
@@ -1870,8 +1878,8 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 			{
 				if(n==S_ESC)
 					c--;
-				if(!lp->lexd.dolparen && (c=sfwrite(sp,bufp,c))>0)
-					iop->iosize += c;
+				if(!lp->lexd.dolparen && (c=(ptrdiff_t)sfwrite(sp,bufp,(size_t)c))>0)
+					iop->iosize += (Sfoff_t)c;
 			}
 			if(LEN==0)
 				SETLEN(1);
@@ -1879,7 +1887,7 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 			{
 				n = LEN;
 				c = fcmbget(&LEN);
-				SETLEN(LEN + n);
+				SETLEN(LEN + (short)n);
 			}
 			else
 				c = lexfill(lp);
@@ -1914,9 +1922,8 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 				if(!lp->lexd.dolparen)
 				{
 					/* write out line */
-					n = fcseek(0)-bufp;
-					if((n=sfwrite(sp,bufp,n))>0)
-						iop->iosize += n;
+					if((n=(ptrdiff_t)sfwrite(sp,bufp,(size_t)(fcseek(0)-bufp)))>0)
+						iop->iosize += (Sfoff_t)n;
 				}
 				/* skip over tabs */
 				if(stripcol)
@@ -1950,8 +1957,8 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 				{
 					if(!lp->lexd.dolparen && (c=cp-bufp))
 					{
-						if((c=sfwrite(sp,cp=bufp,c))>0)
-							iop->iosize+=c;
+						if((c=(ptrdiff_t)sfwrite(sp,cp=bufp,(size_t)c))>0)
+							iop->iosize+=(Sfoff_t)c;
 					}
 					nsave = n;
 					if((c=lexfill(lp))<=0)
@@ -1974,8 +1981,8 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 				{
 					if(!lp->lexd.dolparen && (n=cp-bufp))
 					{
-						if((n=sfwrite(sp,bufp,n))>0)
-							iop->iosize += n;
+						if((n=(ptrdiff_t)sfwrite(sp,bufp,(size_t)n))>0)
+							iop->iosize += (Sfoff_t)n;
 					}
 					sh.inlineno--;
 					if(c==RPAREN)
@@ -1992,8 +1999,8 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 					 */
 					if(!lp->lexd.dolparen && nsave>0)
 					{
-						if((n=sfwrite(sp,iop->iodelim,nsave))>0)
-							iop->iosize += n;
+						if((n=(ptrdiff_t)sfwrite(sp,iop->iodelim,(size_t)nsave))>0)
+							iop->iosize += (Sfoff_t)n;
 						bufp = fcfirst();
 					}
 					if(c==NL)
@@ -2025,8 +2032,8 @@ static int here_copy(Lex_t *lp,struct ionod *iop)
 				sh.inlineno++;
 				if(!lp->lexd.dolparen && (n=(fcseek(0)-bufp)-n)>=0)
 				{
-					if(n && (n=sfwrite(sp,bufp,n))>0)
-						iop->iosize += n;
+					if(n && (n=(ptrdiff_t)sfwrite(sp,bufp,(size_t)n))>0)
+						iop->iosize += (Sfoff_t)n;
 					bufp = fcseek(0)+1;
 				}
 			}
@@ -2063,7 +2070,7 @@ static char	*fmttoken(Lex_t *lp, int sym)
 	if(sym&SYMRES)
 	{
 		const Shtable_t *tp=shtab_reserved;
-		while(tp->sh_number && tp->sh_number!=sym)
+		while(tp->sh_number && (int)tp->sh_number!=sym)
 			tp++;
 		return (char*)tp->sh_name;
 	}
@@ -2155,9 +2162,9 @@ noreturn void sh_syntax(Lex_t *lp, int special)
 static unsigned char *stack_shift(unsigned char *sp, unsigned char *dp)
 {
 	unsigned char *ep;
-	int offset = stktell(sh.stk);
-	int left = offset - (sp - (unsigned char*)stkptr(sh.stk, 0));
-	int shift = (dp+1-sp);
+	ptrdiff_t offset = stktell(sh.stk);
+	ptrdiff_t left = offset - (sp - (unsigned char*)stkptr(sh.stk, 0));
+	ptrdiff_t shift = (dp+1-sp);
 	offset += shift;
 	stkseek(sh.stk,offset);
 	sp = (unsigned char*)stkptr(sh.stk,offset);
@@ -2180,7 +2187,8 @@ static struct argnod *endword(int mode)
 	const char *const state = sh_lexstates[ST_NESTED];
 	unsigned char *sp, *dp, *ep=0, *xp=0;	/* must be unsigned: pointed-to values used as index to 256-byte state table */
 	int inquote=0, inlit=0;			/* set within quoted strings */
-	int n, bracket=0;
+	int bracket=0;
+	ssize_t n;
 	sfputc(sh.stk,0);
 	sp =  (unsigned char*)stkptr(sh.stk,ARGVAL);
 	if(mbwide())
@@ -2267,7 +2275,7 @@ static struct argnod *endword(int mode)
 					}
 					*--dp = 0;
 					msg = ERROR_translate(0,error_info.id,0,ep);
-					n = strlen(msg);
+					n = (ssize_t)strlen(msg);
 					dp = ep+n;
 					if(sp-dp <= 1)
 					{
@@ -2275,7 +2283,7 @@ static struct argnod *endword(int mode)
 						dp = sp-1;
 						ep = dp-n;
 					}
-					memmove(ep,msg,n);
+					memmove(ep,msg,(size_t)n);
 					*dp++ = '"';
 				}
 				ep = 0;
@@ -2335,7 +2343,7 @@ static struct argnod *endword(int mode)
 				if(n=='\n')
 					dp--;
 				else
-					dp[-1] = n;
+					dp[-1] = (unsigned char)n;
 				sp++;
 			}
 			break;
@@ -2455,9 +2463,9 @@ static int alias_exceptf(Sfio_t *iop,int type,void *data, Sfdisc_t *handle)
 	{
 		/* if last character is a blank, then next word can be an alias */
 		int c = fcpeek(-1);
-		if(isblank(c))
+		if(isblank((wint_t)c))
 			lp->aliasok = 1;
-		*ap->buf = ap->nextc;
+		*ap->buf = (char)ap->nextc;
 		ap->nextc = 0;
 		sfsetbuf(iop,ap->buf,1);
 		return 1;
@@ -2508,6 +2516,6 @@ static void setupalias(Lex_t *lp, const char *string,Namval_t *np)
 static int stack_grow(void)
 {
 	lex_max += STACK_ARRAY;
-	lex_match = sh_realloc(lex_match,sizeof(int)*lex_max);
+	lex_match = sh_realloc(lex_match,sizeof(int)*(size_t)lex_max);
 	return 1;
 }
diff --git a/src/cmd/ksh93/sh/parse.c b/src/cmd/ksh93/sh/parse.c
index 7c7dd73cf..d97007f40 100644
--- a/src/cmd/ksh93/sh/parse.c
+++ b/src/cmd/ksh93/sh/parse.c
@@ -37,6 +37,7 @@
 #include	"test.h"
 #include	"history.h"
 #include	"version.h"
+#include	<assert.h>
 
 #define HERE_MEM	SFIO_BUFSIZE	/* size of here-docs kept in memory */
 
@@ -79,8 +80,10 @@ static unsigned long writedefs(Lex_t *lexp,struct argnod *arglist, int line, int
 {
 	struct argnod *argp = arglist;
 	char *cp;
-	int n,eline;
-	int width=0;
+	int eline;
+	size_t width=0;
+	int n;
+	ssize_t len;
 	unsigned long r=0;
 	static char atbuff[20];
 	int  justify=0;
@@ -114,11 +117,11 @@ static unsigned long writedefs(Lex_t *lexp,struct argnod *arglist, int line, int
 			while((n= *++cp))
 			{
 				if(isdigit(n))
-					width = 10*width + n-'0';
+					width = 10*width + (size_t)n-'0';
 				else if(n=='L' || n=='R' || n =='Z')
-					justify=n;
+					justify = n;
 				else
-					*attribute++ = n;
+					*attribute++ = (char)n;
 			}
 		}
 	}
@@ -128,11 +131,11 @@ static unsigned long writedefs(Lex_t *lexp,struct argnod *arglist, int line, int
 	while(argp)
 	{
 		if((cp=strchr(argp->argval,'='))||(cp=strchr(argp->argval,'?')))
-			n = cp-argp->argval;
+			p = (ssize_t)(cp-argp->argval);
 		else
-			n = strlen(argp->argval);
+			p = (ssize_t)strlen(argp->argval);
 		eline = sh.inlineno-(lexp->token==NL);
-		r=kiaentity(lexp,argp->argval,n,type,line,eline,parent,justify,width,atbuff);
+		r=kiaentity(lexp,argp->argval,p,type,line,eline,parent,justify,width,atbuff);
 		sfprintf(kia.tmp,"p;%..64d;v;%..64d;%d;%d;s;\n",kia.current,r,line,eline);
 		argp = argp->argnxt.ap;
 	}
@@ -396,7 +399,7 @@ void	*sh_parse(Sfio_t *iop, int flag)
 	Lex_t		*lexp = (Lex_t*)sh.lex_context;
 	Fcin_t		sav_input;
 	struct argnod	*sav_arg = lexp->arg;
-	int		sav_prompt = sh.nextprompt;
+	short		sav_prompt = sh.nextprompt;
 	if(sh.binscript && (sffileno(iop)==sh.infd || (flag&SH_FUNEVAL)))
 		return sh_trestore(iop);
 	fcsave(&sav_input);
@@ -738,10 +741,10 @@ static struct regnod*	syncase(Lex_t *lexp,int esym)
 static Shnode_t	*arithfor(Lex_t *lexp,Shnode_t *tf)
 {
 	Shnode_t	*t, *tw = tf;
-	int		offset;
+	ptrdiff_t	offset;
 	struct argnod	*argp;
 	int		n;
-	int		argflag = lexp->arg->argflag;
+	uint8_t		argflag = lexp->arg->argflag;
 	Fcin_t		sav_input;
 	/* save current input */
 	fcsave(&sav_input);
@@ -762,7 +765,7 @@ static Shnode_t	*arithfor(Lex_t *lexp,Shnode_t *tf)
 		if(!lexp->token)
 			break;
 		/* remove trailing white space */
-		while(offset>ARGVAL && ((c= *stkptr(sh.stk,offset-1)),isspace(c)))
+		while(offset>(signed)ARGVAL && ((c= *stkptr(sh.stk,offset-1)),isspace(c)))
 			offset--;
 		/* check for empty initialization expression */
 		if(offset==ARGVAL && n==0)
@@ -778,7 +781,7 @@ static Shnode_t	*arithfor(Lex_t *lexp,Shnode_t *tf)
 		else
 			tw->wh.whtre = t;
 	}
-	while((offset=fcpeek(0)) && isspace(offset))
+	while((offset=fcpeek(0)) && isspace((int)offset))
 		fcseek(1);
 	sfputr(sh.stk,fcseek(0),-1);
 	argp = stkfreeze(sh.stk,1);
@@ -819,7 +822,9 @@ static Shnode_t *funct(Lex_t *lexp)
 #if SHOPT_KIA
 	unsigned long current = kia.current;
 #endif /* SHOPT_KIA */
-	int nargs=0,size=0,jmpval;
+	size_t nargs=0;
+	size_t size=0;
+	int jmpval;
 	struct  checkpt buff;
 	int save_optget = opt_get;
 	void	*in_mktype = sh.mktype;
@@ -867,7 +872,7 @@ static Shnode_t *funct(Lex_t *lexp)
 			}
 			if(c)
 				sh_syntax(lexp,2);
-			nargs = argv-argv0;
+			nargs = (size_t)(argv-argv0);
 			size += sizeof(struct dolnod)+(nargs+ARG_SPARE)*sizeof(char*);
 			if(sh.shcomp && strncmp(".sh.math.",t->funct.functnam,9)==0)
 			{
@@ -875,7 +880,7 @@ static Shnode_t *funct(Lex_t *lexp)
 				Namval_t *np= nv_open(t->funct.functnam,sh.fun_tree,NV_ADD|NV_VARNAME);
 				rp = np->nvalue = new_of(struct Ufunction,sh.funload?sizeof(Dtlink_t):0);
 				memset(rp, 0, sizeof(struct Ufunction));
-				rp->argc = ac->comarg.dp->dolnum;
+				rp->argc = (short)ac->comarg.dp->dolnum;
 			}
 		}
 		while(lexp->token==NL)
@@ -997,25 +1002,26 @@ static int check_array(Lex_t *lexp)
 static struct argnod *assign(Lex_t *lexp, struct argnod *ap, int type)
 {
 	int n;
+	size_t len;
 	Shnode_t *t, **tp;
 	struct comnod *ac = NULL;
 	int array=0, index=0;
 	Namval_t *np;
 	lexp->assignlevel++;
-	n = strlen(ap->argval)-1;
-	if(ap->argval[n]!='=')
+	len = strlen(ap->argval)-1;
+	if(ap->argval[len]!='=')
 		sh_syntax(lexp,0);
-	if(ap->argval[n-1]=='+')
+	if(ap->argval[len-1]=='+')
 	{
-		ap->argval[n--]=0;
+		ap->argval[len--]=0;
 		array = ARG_APPEND;
 		type |= NV_APPEND;
 	}
 	/* shift right */
-	while(n > 0)
+	while(len > 0)
 	{
-		ap->argval[n] = ap->argval[n-1];
-		n--;
+		ap->argval[len] = ap->argval[len-1];
+		len--;
 	}
 	*ap->argval=0;
 	t = getnode(fornod);
@@ -1196,7 +1202,8 @@ static Shnode_t	*item(Lex_t *lexp,int flag)
 	int tok = (lexp->token&0xff);
 	int savwdval = lexp->lasttok;
 	int savline = lexp->lastline;
-	int showme=0, comsub;
+	int showme=0;
+	char comsub;
 	if(!(flag&SH_NOIO) && (tok=='<' || tok=='>' || lexp->token==IOVNAME))
 		io=inout(lexp,NULL,1);
 	else
@@ -1399,12 +1406,12 @@ static struct argnod *process_sub(Lex_t *lexp,int tok)
 {
 	struct argnod *argp;
 	Shnode_t *t;
-	int mode = (tok==OPROCSYM);
+	uint8_t mode = (tok==OPROCSYM);
 	t = sh_cmd(lexp,RPAREN,SH_NL);
 	argp = stkalloc(sh.stk,sizeof(struct argnod));
 	*argp->argval = 0;
 	argp->argchn.ap = (struct argnod*)makeparent(lexp,mode?TFORK|FPIN|FAMP|FPCL:TFORK|FPOU,t);
-	argp->argflag =  (ARG_EXP|mode);
+	argp->argflag = (ARG_EXP|mode);
 	return argp;
 }
 
@@ -1424,7 +1431,7 @@ static Shnode_t *simple(Lex_t *lexp,int flag, struct ionod *io)
 	int		was_assign = 0;
 	int		argno = 0;
 	int		assignment = 0;
-	int		key_on = (!(flag&SH_NOIO) && sh_isoption(SH_KEYWORD));
+	char		key_on = (!(flag&SH_NOIO) && sh_isoption(SH_KEYWORD));
 	int		associative=0;
 	if((argp=lexp->arg) && (argp->argflag&ARG_ASSIGN) && argp->argval[0]=='[')
 	{
@@ -1462,7 +1469,7 @@ static Shnode_t *simple(Lex_t *lexp,int flag, struct ionod *io)
 				if(assignment==1)
 				{
 					stkseek(sh.stk,ARGVAL);
-					sfwrite(sh.stk,argp->argval,lexp->varnamelength);
+					sfwrite(sh.stk,argp->argval,(size_t)lexp->varnamelength);
 					ap = stkfreeze(sh.stk,1);
 					ap->argflag = ARG_RAW;
 					ap->argchn.ap = 0;
@@ -1530,7 +1537,7 @@ static Shnode_t *simple(Lex_t *lexp,int flag, struct ionod *io)
 		{
 			if(argp->argflag&ARG_ASSIGN)
 			{
-				int intypeset = lexp->intypeset;
+				char intypeset = lexp->intypeset;
 				lexp->intypeset = 0;
 				if(t->comnamp == SYSCOMPOUND)
 					type = NV_COMVAR;
@@ -1792,7 +1799,7 @@ static struct ionod	*inout(Lex_t *lexp,struct ionod *lastio,int flag)
 		if(lexp->arg->argflag&ARG_RAW)
 			iof |= IORAW;
 	}
-	iop->iofile=iof;
+	iop->iofile=(unsigned)iof;
 	if(flag>0)
 		/* allow alias substitutions and parameter assignments */
 		lexp->aliasok = lexp->assignok = 1;
@@ -1815,6 +1822,7 @@ static struct ionod	*inout(Lex_t *lexp,struct ionod *lastio,int flag)
 		{
 			/* redirect standard output to standard error */
 			ioq = stkalloc(sh.stk,sizeof(struct ionod));
+			assert(ioq != NULL);
 			memset(ioq,0,sizeof(*ioq));
 			ioq->ioname = "1";
 			ioq->iolst = 0;
@@ -1875,7 +1883,7 @@ static struct argnod *qscan(struct comnod *ac,int argn)
 			errormsg(SH_DICT,ERROR_warn(0),message,ac->comline);
 	}
 	/* leave space for an extra argument at the front */
-	dp = stkalloc(sh.stk,(unsigned)sizeof(struct dolnod) + ARG_SPARE*sizeof(char*) + argn*sizeof(char*));
+	dp = stkalloc(sh.stk,sizeof(struct dolnod) + ARG_SPARE*sizeof(char*) + (size_t)argn*sizeof(char*));
 	cp = dp->dolval+ARG_SPARE;
 	dp->dolnum = argn;
 	dp->dolbot = ARG_SPARE;
@@ -2050,13 +2058,13 @@ static Shnode_t *test_primary(Lex_t *lexp)
  * return an entity checksum
  * The entity is created if it doesn't exist
  */
-unsigned long kiaentity(Lex_t *lexp,const char *name,int len,int type,int first,int last,unsigned long parent, int pkind, int width, const char *attr)
+unsigned long kiaentity(Lex_t *lexp,const char *name,ssize_t len,int type,int first,int last,unsigned long parent, int pkind, size_t width, const char *attr)
 {
 	Namval_t *np;
-	long offset = stktell(sh.stk);
+	ptrdiff_t offset = stktell(sh.stk);
 	sfputc(sh.stk,type);
 	if(len>0)
-		sfwrite(sh.stk,name,len);
+		sfwrite(sh.stk,name,(size_t)len);
 	else
 	{
 		if(type=='p')
@@ -2076,9 +2084,9 @@ unsigned long kiaentity(Lex_t *lexp,const char *name,int len,int type,int first,
 		if(!pkind)
 			pkind = '0';
 		if(len>0)
-			sfprintf(kia.file,"%..64d;%c;%.*s;%d;%d;%..64d;%..64d;%c;%d;%s\n",np->hash,type,len,name,first,last,parent,kia.fscript,pkind,width,attr);
+			sfprintf(kia.file,"%..64d;%c;%.*s;%d;%d;%..64d;%..64d;%c;%zu;%s\n",np->hash,type,len,name,first,last,parent,kia.fscript,pkind,width,attr);
 		else
-			sfprintf(kia.file,"%..64d;%c;%s;%d;%d;%..64d;%..64d;%c;%d;%s\n",np->hash,type,name,first,last,parent,kia.fscript,pkind,width,attr);
+			sfprintf(kia.file,"%..64d;%c;%s;%d;%d;%..64d;%..64d;%c;%zu;%s\n",np->hash,type,name,first,last,parent,kia.fscript,pkind,width,attr);
 	}
 	return np->hash;
 }
@@ -2094,25 +2102,25 @@ static void kia_add(Namval_t *np, void *data)
 
 int kiaclose(Lex_t *lexp)
 {
-	off_t off1,off2;
-	int n;
+	Sflong_t off1,off2;
+	ssize_t n;
 	if(kia.file)
 	{
 		unsigned long r = kiaentity(lexp,kia.scriptname,-1,'p',-1,sh.inlineno-1,0,'s',0,"");
 		kiaentity(lexp,kia.scriptname,-1,'p',1,sh.inlineno-1,r,'s',0,"");
 		kiaentity(lexp,kia.scriptname,-1,'f',1,sh.inlineno-1,r,'s',0,"");
 		nv_scan(kia.entity_tree,kia_add,lexp,NV_TAGGED,0);
-		off1 = sfseek(kia.file,0,SEEK_END);
+		off1 = (Sflong_t)sfseek(kia.file,0,SEEK_END);
 		sfseek(kia.tmp,0,SEEK_SET);
 		sfmove(kia.tmp,kia.file,SFIO_UNBOUND,-1);
-		off2 = sfseek(kia.file,0,SEEK_END);
+		off2 = (Sflong_t)sfseek(kia.file,0,SEEK_END);
 		if(off2==off1)
-			n= sfprintf(kia.file,"DIRECTORY\nENTITY;%jd;%zu\nDIRECTORY;",(Sflong_t)kia.begin,(size_t)(off1-kia.begin));
+			n= sfprintf(kia.file,"DIRECTORY\nENTITY;%jd;%jd\nDIRECTORY;",(Sflong_t)kia.begin,off1-(Sflong_t)kia.begin);
 		else
-			n= sfprintf(kia.file,"DIRECTORY\nENTITY;%jd;%zu\nRELATIONSHIP;%jd;%zu\nDIRECTORY;",(Sflong_t)kia.begin,(size_t)(off1-kia.begin),(Sflong_t)off1,(size_t)(off2-off1));
+			n= sfprintf(kia.file,"DIRECTORY\nENTITY;%jd;%jd\nRELATIONSHIP;%jd;%jd\nDIRECTORY;",(Sflong_t)kia.begin,off1-(Sflong_t)kia.begin,off1,off2-off1);
 		if(off2 >= INT_MAX)
 			off2 = -(n+12);
-		sfprintf(kia.file,"%010.10jd;%010d\n",(Sflong_t)off2+10, n+12);
+		sfprintf(kia.file,"%010.10jd;%010zd\n",off2+10, n+12);
 	}
 	return sfclose(kia.file);
 }
diff --git a/src/cmd/ksh93/sh/shcomp.c b/src/cmd/ksh93/sh/shcomp.c
index 6016d46d4..c3eb8ca51 100644
--- a/src/cmd/ksh93/sh/shcomp.c
+++ b/src/cmd/ksh93/sh/shcomp.c
@@ -140,7 +140,7 @@ int main(int argc, char *argv[])
 			UNREACHABLE();
 		}
 		if(fstat(sffileno(out),&statb) >=0)
-			chmod(cp,(statb.st_mode&~S_IFMT)|S_IXUSR|S_IXGRP|S_IXOTH);
+			chmod(cp,(statb.st_mode&(mode_t)~S_IFMT)|S_IXUSR|S_IXGRP|S_IXOTH);
 	}
 	else
 		out = sfstdout;
diff --git a/src/cmd/ksh93/sh/subshell.c b/src/cmd/ksh93/sh/subshell.c
index dff9ee374..df1dbe832 100644
--- a/src/cmd/ksh93/sh/subshell.c
+++ b/src/cmd/ksh93/sh/subshell.c
@@ -68,7 +68,7 @@ static struct subshell
 	int		tmpfd;	/* saved tmp file descriptor */
 	int		pipefd;	/* read fd if pipe is created */
 	char		jobcontrol;
-	unsigned char	fdstatus;
+	uint8_t		fdstatus;
 	int		fdsaved; /* bit mask for saved file descriptors */
 	int		sig;	/* signal for $$ */
 	pid_t		bckpid;
@@ -316,7 +316,8 @@ static void nv_restore(struct subshell *sp)
 	struct Link	*lp, *lq;
 	Namval_t	*mp, *np;
 	Namval_t	*mpnext;
-	int		flags,nofree;
+	int		flags;
+	char		nofree;
 	sh.nv_restore = 1;
 	for(lp=sp->svar; lp; lp=lq)
 	{
@@ -521,7 +522,7 @@ void sh_clear_subshell_pwdfd(void)
  * If comsub is not null, the return value will be a stream consisting of
  * output of command <t>.  Otherwise, NULL will be returned.
  */
-Sfio_t *sh_subshell(Shnode_t *t, volatile int flags, int comsub)
+Sfio_t *sh_subshell(Shnode_t *t, volatile int flags, char comsub)
 {
 	struct subshell sub_data;
 	struct subshell *sp = &sub_data;
@@ -728,7 +729,8 @@ Sfio_t *sh_subshell(Shnode_t *t, volatile int flags, int comsub)
 		{
 			if(sh.spid)
 			{
-				int e = sh.exitval, c = sh.chldexitsig;
+				int e = sh.exitval;
+				char c = sh.chldexitsig;
 				job_wait(sh.spid);
 				sh.exitval = e, sh.chldexitsig = c;
 				if(sh.pipepid==sh.spid)
@@ -842,7 +844,7 @@ Sfio_t *sh_subshell(Shnode_t *t, volatile int flags, int comsub)
 		n = sh.st.trapmax-savst.trapmax;
 		sh_sigreset(1);
 		if(n>0)
-			memset(&sh.st.trapcom[savst.trapmax],0,n*sizeof(char*));
+			memset(&sh.st.trapcom[savst.trapmax],0,(size_t)n*sizeof(char*));
 		sh.st = savst;
 		sh.st.otrap = 0;
 		if(nsig)
diff --git a/src/cmd/ksh93/sh/tdump.c b/src/cmd/ksh93/sh/tdump.c
index 19da2f796..1db7b1854 100644
--- a/src/cmd/ksh93/sh/tdump.c
+++ b/src/cmd/ksh93/sh/tdump.c
@@ -2,7 +2,7 @@
 *                                                                      *
 *               This software is part of the ast package               *
 *          Copyright (c) 1982-2011 AT&T Intellectual Property          *
-*          Copyright (c) 2020-2025 Contributors to ksh 93u+m           *
+*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
 *                      and is licensed under the                       *
 *                 Eclipse Public License, Version 2.0                  *
 *                                                                      *
@@ -29,17 +29,17 @@
 #include	"path.h"
 #include	"io.h"
 
-static int p_comlist(const struct dolnod*);
-static int p_arg(const struct argnod*);
-static int p_comarg(const struct comnod*);
-static int p_redirect(const struct ionod*);
-static int p_switch(const struct regnod*);
-static int p_tree(const Shnode_t*);
-static int p_string(const char*);
+static ssize_t p_comlist(const struct dolnod*);
+static ssize_t p_arg(const struct argnod*);
+static ssize_t p_comarg(const struct comnod*);
+static ssize_t p_redirect(const struct ionod*);
+static ssize_t p_switch(const struct regnod*);
+static ssize_t p_tree(const Shnode_t*);
+static ssize_t   p_string(const char*);
 
 static Sfio_t *outfile;
 
-int sh_tdump(Sfio_t *out, const Shnode_t *t)
+ssize_t sh_tdump(Sfio_t *out, const Shnode_t *t)
 {
 	outfile = out;
 	return p_tree(t);
@@ -48,7 +48,7 @@ int sh_tdump(Sfio_t *out, const Shnode_t *t)
 /*
  * print script corresponding to shell tree <t>
  */
-static int p_tree(const Shnode_t *t)
+static ssize_t p_tree(const Shnode_t *t)
 {
 	if(!t)
 		return sfputl(outfile,-1);
@@ -135,9 +135,9 @@ static int p_tree(const Shnode_t *t)
 	return -1;
 }
 
-static int p_arg(const struct argnod *arg)
+static ssize_t p_arg(const struct argnod *arg)
 {
-	int n;
+	size_t n;
 	struct fornod *fp;
 	while(arg)
 	{
@@ -169,7 +169,7 @@ static int p_arg(const struct argnod *arg)
 	return sfputu(outfile,0);
 }
 
-static int p_redirect(const struct ionod *iop)
+static ssize_t p_redirect(const struct ionod *iop)
 {
 	while(iop)
 	{
@@ -184,7 +184,7 @@ static int p_redirect(const struct ionod *iop)
 		if(iop->iodelim)
 		{
 			p_string(iop->iodelim);
-			sfputl(outfile,iop->iosize);
+			sfputl(outfile,(Sflong_t)iop->iosize);
 			sfseek(sh.heredocs,iop->iooffset,SEEK_SET);
 			sfmove(sh.heredocs,outfile, iop->iosize,-1);
 		}
@@ -197,7 +197,7 @@ static int p_redirect(const struct ionod *iop)
 	return sfputl(outfile,-1);
 }
 
-static int p_comarg(const struct comnod *com)
+static ssize_t p_comarg(const struct comnod *com)
 {
 	p_redirect(com->comio);
 	p_arg(com->comset);
@@ -210,10 +210,10 @@ static int p_comarg(const struct comnod *com)
 	return sfputu(outfile,com->comline);
 }
 
-static int p_comlist(const struct dolnod *dol)
+static ssize_t p_comlist(const struct dolnod *dol)
 {
 	char *cp, *const*argv;
-	int n;
+	ptrdiff_t n;
 	argv = dol->dolval+ARG_SPARE;
 	while(cp = *argv)
 		argv++;
@@ -225,7 +225,7 @@ static int p_comlist(const struct dolnod *dol)
 	return sfputu(outfile,0);
 }
 
-static int p_switch(const struct regnod *reg)
+static ssize_t p_switch(const struct regnod *reg)
 {
 	while(reg)
 	{
@@ -237,7 +237,7 @@ static int p_switch(const struct regnod *reg)
 	return sfputl(outfile,-1);
 }
 
-static int p_string(const char *string)
+static ssize_t p_string(const char *string)
 {
 	size_t n=strlen(string);
 	if(sfputu(outfile,n+1)<0)
diff --git a/src/cmd/ksh93/sh/timers.c b/src/cmd/ksh93/sh/timers.c
index 26e96d854..18552cee7 100644
--- a/src/cmd/ksh93/sh/timers.c
+++ b/src/cmd/ksh93/sh/timers.c
@@ -2,7 +2,7 @@
 *                                                                      *
 *               This software is part of the ast package               *
 *          Copyright (c) 1982-2012 AT&T Intellectual Property          *
-*          Copyright (c) 2020-2024 Contributors to ksh 93u+m           *
+*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
 *                      and is licensed under the                       *
 *                 Eclipse Public License, Version 2.0                  *
 *                                                                      *
@@ -200,7 +200,7 @@ void *sh_timeradd(Sfulong_t msec,int flags,void (*action)(void*),void *handle)
 		{
 			Handler_t *hp = (Handler_t*)sh_malloc(sizeof(Handler_t));
 			*hp = fn;
-			sh_timeradd((Sflong_t)(1000*t), 0, oldalrm, hp);
+			sh_timeradd((Sfulong_t)(1000*t), 0, oldalrm, hp);
 		}
 		tp = tptop;
 	}
diff --git a/src/cmd/ksh93/sh/trestore.c b/src/cmd/ksh93/sh/trestore.c
index 8911b507e..33532ca1c 100644
--- a/src/cmd/ksh93/sh/trestore.c
+++ b/src/cmd/ksh93/sh/trestore.c
@@ -2,7 +2,7 @@
 *                                                                      *
 *               This software is part of the ast package               *
 *          Copyright (c) 1982-2011 AT&T Intellectual Property          *
-*          Copyright (c) 2020-2024 Contributors to ksh 93u+m           *
+*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
 *                      and is licensed under the                       *
 *                 Eclipse Public License, Version 2.0                  *
 *                                                                      *
@@ -51,12 +51,12 @@ Shnode_t *sh_trestore(Sfio_t *in)
  */
 static Shnode_t *r_tree(void)
 {
-	long l = sfgetl(infile);
+	Sflong_t l = sfgetl(infile);
 	int type;
 	Shnode_t *t=0;
 	if(l<0)
 		return t;
-	type = l;
+	type = (int)l;
 	switch(type&COMMSK)
 	{
 		case TTIME:
@@ -72,7 +72,7 @@ static Shnode_t *r_tree(void)
 		case TSETIO:
 		case TFORK:
 			t = getnode(forknod);
-			t->fork.forkline = sfgetu(infile);
+			t->fork.forkline = (int)sfgetu(infile);
 			t->fork.forktre = r_tree();
 			t->fork.forkio = r_redirect();
 			break;
@@ -98,7 +98,7 @@ static Shnode_t *r_tree(void)
 			break;
 		case TARITH:
 			t = getnode(arithnod);
-			t->ar.arline = sfgetu(infile);
+			t->ar.arline = (int)sfgetu(infile);
 			t->ar.arexpr = r_arg();
 			t->ar.arcomp = 0;
 			if((t->ar.arexpr)->argflag&ARG_RAW)
@@ -108,7 +108,7 @@ static Shnode_t *r_tree(void)
 			t = getnode(fornod);
 			t->for_.forline = 0;
 			if(type&FLINENO)
-				t->for_.forline = sfgetu(infile);
+				t->for_.forline = (int)sfgetu(infile);
 			t->for_.fortre = r_tree();
 			t->for_.fornam = r_string();
 			t->for_.forlst = (struct comnod*)r_tree();
@@ -117,7 +117,7 @@ static Shnode_t *r_tree(void)
 			t = getnode(swnod);
 			t->sw.swline = 0;
 			if(type&FLINENO)
-				t->sw.swline = sfgetu(infile);
+				t->sw.swline = (int)sfgetu(infile);
 			t->sw.swarg = r_arg();
 			if(type&COMSCAN)
 				t->sw.swio = r_redirect();
@@ -131,7 +131,7 @@ static Shnode_t *r_tree(void)
 			struct slnod *slp;
 			struct functnod *fp;
 			t = getnode(functnod);
-			t->funct.functline = sfgetu(infile);
+			t->funct.functline = (int)sfgetu(infile);
 			t->funct.functnam = r_string();
 			savstak = sh.stk;
 			sh.stk = stkopen(STK_SMALL);
@@ -154,7 +154,7 @@ static Shnode_t *r_tree(void)
 		}
 		case TTST:
 			t = getnode(tstnod);
-			t->tst.tstline = sfgetu(infile);
+			t->tst.tstline = (int)sfgetu(infile);
 			if((type&TPAREN)==TPAREN)
 				t->lst.lstlef = r_tree();
 			else
@@ -172,11 +172,11 @@ static Shnode_t *r_tree(void)
 static struct argnod *r_arg(void)
 {
 	struct argnod	*ap=0, *apold, *aptop=0;
-	long		l;
+	Sfulong_t	l;
 	Stk_t		*stkp=sh.stk;
 	while((l=sfgetu(infile))>0)
 	{
-		ap = stkseek(stkp,(unsigned)l+ARGVAL);
+		ap = stkseek(stkp,(ptrdiff_t)l+(ptrdiff_t)ARGVAL);
 		if(!aptop)
 			aptop = ap;
 		else
@@ -185,14 +185,14 @@ static struct argnod *r_arg(void)
 			sfread(infile,ap->argval,(size_t)l);
 		ap->argval[l] = 0;
 		ap->argchn.cp = 0;
-		ap->argflag = sfgetc(infile);
+		ap->argflag = (uint8_t)sfgetc(infile);
 		ap = stkfreeze(stkp,0);
 		if(*ap->argval==0 && (ap->argflag&ARG_EXP))
 			ap->argchn.ap = (struct argnod*)r_tree();
 		else if(*ap->argval==0 && (ap->argflag&~(ARG_APPEND|ARG_MESSAGE|ARG_QUOTED|ARG_ARRAY))==0)
 		{
 			struct fornod *fp = (struct fornod*)getnode(fornod);
-			fp->fortyp = sfgetu(infile);
+			fp->fortyp = (int)sfgetu(infile);
 			fp->fortre = r_tree();
 			fp->fornam = ap->argval+1;
 			ap->argchn.ap = (struct argnod*)fp;
@@ -206,7 +206,7 @@ static struct argnod *r_arg(void)
 
 static struct ionod *r_redirect(void)
 {
-	long l;
+	Sflong_t l;
 	struct ionod *iop=0, *iopold, *ioptop=0;
 	while((l=sfgetl(infile))>=0)
 	{
@@ -215,14 +215,14 @@ static struct ionod *r_redirect(void)
 			ioptop = iop;
 		else
 			iopold->ionxt = iop;
-		iop->iofile = l;
+		iop->iofile = (unsigned int)l;
 		if((l & IOPROCSUB) && !(l & IOLSEEK))
 			iop->ioname = (char*)r_tree();	/* process substitution as file name to redirection */
 		else
 			iop->ioname = r_string();	/* file name, descriptor, etc. */
 		if(iop->iodelim = r_string())
 		{
-			iop->iosize = sfgetl(infile);
+			iop->iosize = (Sfoff_t)sfgetl(infile);
 			if(sh.heredocs)
 				iop->iooffset = sfseek(sh.heredocs,0,SEEK_END);
 			else
@@ -237,7 +237,7 @@ static struct ionod *r_redirect(void)
 			iop->iovname = r_string();
 		else
 			iop->iovname = 0;
-		iop->iofile &= ~IOVNM;
+		iop->iofile &= (unsigned)~IOVNM;
 	}
 	if(iop)
 		iop->ionxt = 0;
@@ -258,7 +258,7 @@ static void r_comarg(struct comnod *com)
 	}
 	else if(com->comarg.dp = r_comlist())
 		cmdname = com->comarg.dp->dolval[ARG_SPARE];
-	com->comline = sfgetu(infile);
+	com->comline = (int)sfgetu(infile);
 	com->comnamq = 0;
 	if(cmdname)
 	{
@@ -278,12 +278,12 @@ static void r_comarg(struct comnod *com)
 static struct dolnod *r_comlist(void)
 {
 	struct dolnod *dol=0;
-	long l;
+	Sflong_t l;
 	char **argv;
 	if((l=sfgetl(infile))>0)
 	{
-		dol = stkalloc(sh.stk,sizeof(struct dolnod) + sizeof(char*)*(l+ARG_SPARE));
-		dol->dolnum = l;
+		dol = stkalloc(sh.stk,sizeof(struct dolnod) + sizeof(char*)*(size_t)(l+ARG_SPARE));
+		dol->dolnum = (int)l;
 		dol->dolbot = ARG_SPARE;
 		argv = dol->dolval+ARG_SPARE;
 		while(*argv++ = r_string());
@@ -293,7 +293,7 @@ static struct dolnod *r_comlist(void)
 
 static struct regnod *r_switch(void)
 {
-	long l;
+	Sflong_t l;
 	struct regnod *reg=0,*regold,*regtop=0;
 	while((l=sfgetl(infile))>=0)
 	{
@@ -302,7 +302,7 @@ static struct regnod *r_switch(void)
 			regtop = reg;
 		else
 			regold->regnxt = reg;
-		reg->regflag = l;
+		reg->regflag = (char)l;
 		reg->regptr = r_arg();
 		reg->regcom = r_tree();
 		regold = reg;
@@ -315,12 +315,12 @@ static struct regnod *r_switch(void)
 static char *r_string(void)
 {
 	Sfio_t *in = infile;
-	unsigned long l = sfgetu(in);
+	size_t l = (size_t)sfgetu(in);
 	char *ptr;
 	if(l == 0)
 		return NULL;
-	ptr = stkalloc(sh.stk,(unsigned)l);
-	if(--l > 0 && sfread(in,ptr,(size_t)l) != (size_t)l)
+	ptr = stkalloc(sh.stk,l);
+	if(--l > 0 && (size_t)sfread(in,ptr,l) != l)
 		return NULL;
 	ptr[l] = 0;
 	return ptr;
-- 
2.53.0

